content/5_Plat/Sparse_Segtree.mdx

```diff
@@ -63,130 +62,212 @@ range is up to $10^9$, so we have to use a sparse segment tree.
+		int m = (l + r) / 2;
+		apply(cur->left, m - l + 1, cur->lazy);
+		apply(cur->right, r - m, cur->lazy);
+	}
 
-const int MAXN = 123456;
-Node segtree[64 * MAXN];
-int cnt = 2;
-
-void push_lazy(int node) {
-	if (segtree[node].lazy) {
-		segtree[node].sum = segtree[node].tr - segtree[node].tl + 1;
-		int mid = (segtree[node].tl + segtree[node].tr) / 2;
-		if (segtree[node].l == -1) {
-			segtree[node].l = cnt++;
-			segtree[segtree[node].l].tl = segtree[node].tl;
-			segtree[segtree[node].l].tr = mid;
+	void range_set(Node *cur, int l, int r, int ql, int qr, int val) {
+		if (qr < l || ql > r) { return; }
+		if (ql <= l && r <= qr) {
+			apply(cur, r - l + 1, val);
+		} else {
+			push_down(cur, l, r);
+			int m = (l + r) / 2;
+			range_set(cur->left, l, m, ql, qr, val);
+			range_set(cur->right, m + 1, r, ql, qr, val);
+			(cur->freq) = comb((cur->left)->freq, (cur->right)->freq);
 		}
-		if (segtree[node].r == -1) {
-			segtree[node].r = cnt++;
-			segtree[segtree[node].r].tl = mid + 1;
-			segtree[segtree[node].r].tr = segtree[node].tr;
+	}
+
+	int range_sum(Node *cur, int l, int r, int ql, int qr) {
+		if (qr < l || ql > r) { return 0; }
+		if (ql <= l && r <= qr) { return cur->freq; }
+		push_down(cur, l, r);
+		int m = (l + r) / 2;
+		return comb(range_sum(cur->left, l, m, ql, qr),
+		            range_sum(cur->right, m + 1, r, ql, qr));
+	}
+
+  public:
+	SparseSegtree(int n) : n(n) {}
+
+	void range_set(int ql, int qr, int val) { range_set(root, 0, n - 1, ql, qr, val); }
+
+	int range_sum(int ql, int qr) { return range_sum(root, 0, n - 1, ql, qr); }
+};
+
+int main() {
+	int query_num;
+	cin >> query_num;
+	const int RANGE_SIZE = 1e9;
+	SparseSegtree st(RANGE_SIZE + 1);
+
+	int c = 0;
+	for (int i = 0; i < query_num; i++) {
+		int type, x, y;
+		cin >> type >> x >> y;
+		if (type == 1) {
+			c = st.range_sum(x + c, y + c);
+			cout << c << '\n';
+		} else if (type == 2) {
+			st.range_set(x + c, y + c, 1);
 		}
-		segtree[segtree[node].l].lazy = segtree[segtree[node].r].lazy = 1;
-		segtree[node].lazy = 0;
 	}
 }
+```
 
-void update(int node, int l, int r) {
-	push_lazy(node);
-	if (l == segtree[node].tl && r == segtree[node].tr) {
-		segtree[node].lazy = 1;
-		push_lazy(node);
-	} else {
-		int mid = (segtree[node].tl + segtree[node].tr) / 2;
-		if (segtree[node].l == -1) {
-			segtree[node].l = cnt++;
-			segtree[segtree[node].l].tl = segtree[node].tl;
-			segtree[segtree[node].l].tr = mid;
-		}
-		if (segtree[node].r == -1) {
-			segtree[node].r = cnt++;
-			segtree[segtree[node].r].tl = mid + 1;
-			segtree[segtree[node].r].tr = segtree[node].tr;
-		}
+<Warning>
 
-		if (l > mid) update(segtree[node].r, l, r);
-		else if (r <= mid) update(segtree[node].l, l, r);
-		else {
-			update(segtree[node].l, l, mid);
-			update(segtree[node].r, mid + 1, r);
-		}
+Instantiating multiple pointer segment trees could result in a memory leak
+because C++ doesn't have a garbage collector.
```

Comment:
i mean even instantiating one is kinda a memory leak since you don't do any cleaning up

Reply:
yeah but it's only problematic if ur running thru multiple test cases and use multiple segtrees

idk, it's typically not an issue but something good to note imo

Reply:
hm hb "note that this impl isn't gc'd, so if you redclare stuff it's kinda cooked"

Reply:
k